lib/influx.rb

```diff
@@ -234,6 +234,90 @@ def threshold_recommendations(opts)
+      return [] if incidents.empty?
+      unacked = []
+      unresolved = []
+      # FIXME: this reject should be done at the time of querying InfluxDB.
+      incidents.reject! {|incident| incident['incident_key'].nil? || incident['time_to_resolve']}
+      incidents = incidents.map do |incident|
+        entity, check = incident['incident_key'].split(':', 2)
+        {
+          'alert_time' => incident["time"],
+          'id' => incident['id'],
+          'entity' => entity,
+          'check'  => check,
+          'ack_by' => incident['acknowledge_by'] || 'N/A',
+          'time_to_ack' => incident['time_to_ack'],
+        }
+      end
+      acked, unacked = incidents.partition {|x| x["time_to_ack"]}
+      [acked, unacked]
+    end
+
+    def generate_stats
+      shifts = TOML.load_file('config.toml')['shifts'] || {}
+      Time.zone = shifts['time_zone'] || 'UTC'
+      Chronic.time_class = Time.zone
+      shifts.delete('time_zone')
+
+      stat_matrix = [
+        { title: "Last 24 hours", start: Chronic.parse('24 hours ago'), end: Chronic.parse('now')},
+        { title: "Last month", start: Chronic.parse('1 month ago'), end: Chronic.parse('now')},
+        { title: "Last 3 months", start: Chronic.parse('3 months ago'), end: Chronic.parse('now')}
+      ]
+
+      # We want to show data for one full shift each, plus the current shift.
+      # Due to timezones, this sometimes means we need to look back over the last 3 days
+      shift_matrix = []
+      [
+        Date.today.prev_day.prev_day.strftime('%F'),
+        Date.today.prev_day.strftime('%F'),
+        Date.today.strftime('%F')
+      ].each do |day|
+        shift_times = shifts.each do |_, shift|
+          start_time = shift['start_time']
+          duration = shift['duration'].to_i
+          start_date = Chronic.parse("#{day} #{start_time}:00")
+          # Eliminate any shifts starting in the future
+          next if start_date.to_i > Time.now.to_i
+          end_date = start_date + duration * 60
+          shift_matrix << { title: shift['name'], start: start_date, end: end_date }
+        end
+      end
+      # Now we remove any full shifts that are older than the most recent full shift
+      num_shifts = shifts.length + 1
+      stat_matrix << shift_matrix.sort_by{ |x| x[:start] }.reverse.take(num_shifts).reverse
+      stat_matrix.flatten!
+
+      aggregate_select_str = "select count(incident_key), " + %w(ack resolve).map { |type|
+        "MEAN(time_to_#{type}) as #{type}_mean, STDDEV(time_to_#{type}) as #{type}_stddev, PERCENTILE(time_to_#{type}, 95) as #{type}_95_percentile"
+      }.join(', ')
+
+      stat_matrix.each do |shift|
+        aggregated = find_incidents(shift[:start], shift[:end],
+                                    :query_select => aggregate_select_str).first || {}
+        ack_percent_in_60s = ack_percent_before_timeout(:start_date => shift[:start], :end_date => shift[:end],
```

Comment:
If the line break here was before `:end_date` the line would be ~86 chars - that seems more reaonable